### PR TITLE
Improve tty error logging when buildkit vertex is unknown

### LINKED ISSUE #2044: 🐞  dagger do doesn't output stderr
### What is the issue?

`dagger do <action>` doesn't return useful info from stderr.

When I change the logging format to `--logformat plain` I get the info I need.

### Log output

```
➜  service-foo git:(master) ✗ dagger do typescript
[✔] client.env                                                                                                                                                       0.0s
[✔] client.filesystem."./".read                                                                                                                                      0.5s
[✗] actions.typescript                                                                                                                                             216.1s
                                                                                                                                                                   209.9s
8:14PM FTL failed to execute plan: task failed: actions.typescript._build: process "/bin/sh -c cp -r $DEPLOYABLE_NAME/build /output" did not complete successfully: exit code: 1
```

```
dagger do typescript --log-format plain

7:51PM INF system | #10 97.11 Failed to compile.
7:51PM INF system | #10 97.11
7:51PM INF system | #10 97.11 src/store/foo.ts
7:51PM INF system | #10 97.11   Line 29:7:  Unsafe return of an `any` typed value  @typescript-eslint/no-unsafe-return
7:51PM INF system | #10 97.1
7:51PM INF system | #10 97.39 Output:
7:51PM INF system | #10 97.39
7:51PM INF system | #10 97.39 info Visit https://yarnpkg.com/en/docs/cli/workspaces for documentation about this command.
7:51PM INF system | #10 97.42 error Command failed with exit code 1.
7:51PM INF system | #10 97.42 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
7:51PM ERR actions.typescript._build | failed: process "/bin/sh -c yarn build" did not complete successfully: exit code: 1
```

### Steps to reproduce

_No response_

### Dagger version

dagger 0.2.4 (b32c8732) darwin/amd64

### OS version

osx - 12.01

### PR DESCRIPTION
**Pull request by marcosnils**
Creates a generic "system" group in the tty output similarly to the plain one which captures
buildkit events that report a non-dagger vertex name. This happens
currently when using core.#Dockerfile actions since Dagger delegates the
LLB generation to buildkit through it's frontend and we don't get
meaningful events that we can correlate from Dagger's side. 

Given the following dockerfile:

```Dockerfile
FROM node:10-alpine
RUN apk update
RUN some_error
```

Old output :

```
dagger do hello
[✔] client.filesystem.".".read                                                                         0.0s
[✗] actions.hello                                                                                      6.9s
                                                                                                       0.8s
5:18PM FTL failed to execute plan: task failed: actions.hello._build: process "/bin/sh -c some_error" did not complete successfully: exit code: 127
```

New output:

```
dagger do  hello
[✔] client.filesystem.".".read                                                                         0.1s
[✗] actions.hello                                                                                      1.8s
#5 0.083 fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
#5 0.605 fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
#5 0.858 v3.11.13-8-gaf7d80ff31 [http://dl-cdn.alpinelinux.org/alpine/v3.11/main]
#5 0.858 v3.11.11-124-gf2729ece5a [http://dl-cdn.alpinelinux.org/alpine/v3.11/community]
#5 0.858 OK: 11280 distinct packages available
#6 0.052 /bin/sh: some_error: not found
5:16PM FTL failed to execute plan: task failed: actions.hello._build: process "/bin/sh -c some_error" did not complete successfully: exit code: 127
```

cc @aluzzardi @helderco 

Fixes #2044 
Related #613

### CODE REVIEW

File: plan/runner.go
Comment (aluzzardi) [inline plan/runner.go:134]: I don't get this part -- if I understand correctly, for Dockerfile tasks, we're logging things twice -- once with `task = ""` and once again with `task = "system"`, then on the tty/plain writers we rewrite `"" -> "system"`.

I don't understand why this is needed?

Comment (marcosnils) [reply plan/runner.go:134]: At this point we're not logging with `task = ""` since we're inside the cueflow taskFunc and we have the correct Dagger task names. 

We're logging twice because otherwise, the task action name won't be printed to the tty. So, for `dockerfile` actions, at this point we need to log twice so both the dockerfile action and the `system` logger get recorded by the tty output.

Comment (marcosnils) [reply plan/runner.go:134]: > Sorry, I'm slow today, still not following :)

np haha! It's likely that I'm missing something here as well. 

> The second log call doesn't contain the task name, right? I don't see why it would change what's printed.

Correct, but we need the second log so the `system` group is state updated and handled by the tty output here: https://github.com/dagger/dagger/blob/main/cmd/dagger/logger/tty.go#L262

> I don't understand the goal of the system entry since we can't rely on computing/completed etc (as you can have a bunch of Dockerfile running in parallel)

You're seeing this in the `plain` output, but the `tty` output needs this state transition so it can correctly display if something failed or not. In the case you have multiple DAG nodes with Dockerfile actions, `system` will have the state of the last execution.

Comment (aluzzardi) [reply plan/runner.go:134]: Got it!

I think there are some edge cases as the "last" execution is fuzzy: you can have N Dockerfile running in parallel, e.g. the state changes will not be linear (#1 computing / #2 computing / #2 completed / #1 completed), so tty will "flip" between computing/completed for the same docker build.

However, I don't know if this is really a big problem in practice and since this PR is a giant improvement already over what we have, I vote we merge it as is and see how it goes :)

**Review by aluzzardi — APPROVED** [review]

File: plan/runner.go
Comment (marcosnils) [reply plan/runner.go:134]: Let's not merge this yet. I want to explore something to see if it can be further improved before merging

**Comment by marcosnils** [comment]
@aluzzardi added a new commit here that improves the way we're handling state transitions for the new `system` output. 

Once thing that I've noticed and spoke with @helderco  is that if you have a Dagger plan which contains parallel tasks, if one fails the others gets cancelled. Here's an example: 

--- test.cue ---
```
package main

import (
	"dagger.io/dagger"
	"universe.dagger.io/bash"
	"universe.dagger.io/alpine"
)

dagger.#Plan & {
	actions: {
		_image: alpine.#Build & {
			packages: {
				bash: _
				npm:  _
			}
		}

		// Hello World Dagger
		hello: {
			hello1: bash.#Run & {
				input: _image.output
				script: contents: #"fail"#
			}

			hello2: bash.#Run & {
				input: _image.output
				script: contents: #"sleep 5;"#
			}
		}

	}
}
``` 

Here's the output that I get when running `dagger do hello`

image

In the case of `Dockerfile` actions which share the same `system` group, my second commit in this PR prevents the `cancelled` action state to override the `failed`, and errors will be shown correctly. 

Below is an example with the `Dockerfile` scenario:

-- Dockerfile.fail ```
``` 
FROM node:10-alpine

RUN fails
```

-- Dockerfile.long --
```
FROM node:10-alpine

RUN sleep 5
```

-- test.cue --
```
package main

import (
	"dagger.io/dagger"
	"universe.dagger.io/docker"
)

dagger.#Plan & {
	client: filesystem: ".": read: contents: dagger.#FS
	actions: {

		// Hello World Dagger
		hello: {
			hello1: docker.#Dockerfile & {
				source: client.filesystem.".".read.contents
                                dockerfile: path: "./Dockerfile.fail"
			}
			hello2: docker.#Dockerfile & {
				source: client.filesystem.".".read.contents
				dockerfile: path: "./Dockerfile.long"
			}

		}

	}
}
```

Before the 2nd commit in this PR output was:

image


^ no output was generated here since "cancelled" is overriding the "failed" state for the "system" group


After the 2nd commit, output now is :

image


^ error is properly shown upon cancellation.

**Comment by aluzzardi** [comment]
LGTM!

Test failure is unrelated -- restarted CI

**Pull request merged**
